### PR TITLE
[JENKINS-43055] Support FlowExecutionListener

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-43055](https://issues.jenkins-ci.org/browse/JENKINS-43055)

Replaces https://github.com/jenkinsci/workflow-job-plugin/pull/50 and https://github.com/jenkinsci/workflow-cps-plugin/pull/127, downstream of https://github.com/jenkinsci/workflow-api-plugin/pull/36

cc @reviewbybees

### CODE REVIEW

File: src/test/java/org/jenkinsci/plugins/workflow/job/WorkflowRunRestartTest.java [hunk 1]
```diff
@@ -175,4 +184,96 @@ public DescriptorImpl() {
...
+                SemaphoreStep.success("wait/1", null);
+
+                story.j.assertBuildStatus(Result.FAILURE, story.j.waitForCompletion(b));
+                story.j.assertLogContains("Running for listener", b);
+
+                ExecListener listener = ExtensionList.lookup(FlowExecutionListener.class).get(ExecListener.class);
+                assertNotNull(listener);
+                assertEquals(1, listener.started);
+                assertEquals(1, listener.resumed);
+                assertEquals(1, listener.finished);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/job/WorkflowRunRestartTest.java]: Could add an extra `semaphore` to demonstrate started & resumed & !finished.

File: src/test/java/org/jenkinsci/plugins/workflow/job/WorkflowRunRestartTest.java [hunk 2]
```diff
@@ -175,4 +184,96 @@ public DescriptorImpl() {
...
+                WorkflowRun b = p.getLastBuild();
+                assertTrue(b.isBuilding());
+                SemaphoreStep.success("wait/1", null);
+
+                story.j.assertBuildStatus(Result.FAILURE, story.j.waitForCompletion(b));
+                story.j.assertLogContains("Running for listener", b);
+
+                ExecListener listener = ExtensionList.lookup(FlowExecutionListener.class).get(ExecListener.class);
+                assertNotNull(listener);
+                assertEquals(1, listener.started);
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/job/WorkflowRunRestartTest.java]: Potentially confusing to readers. This `started` is not the same as the one in the first step: `onRunning(…, true)` increments _both_ fields, but we are using a new listener!

This is why I wonder whether having a single method overloaded to two events is wise.

Comment (abayer) [inline src/test/java/org/jenkinsci/plugins/workflow/job/WorkflowRunRestartTest.java]: Yeaaaah, it's not. Fixing.

**Comment by abayer** [comment]
:bee: Hold off on this for the moment - something's wrong in terms of the `SyntheticStageGraphListener` attaching in Declarative.

**Comment by abayer** [comment]
Ok, findbugs fixed - @jglick, back to you for final review/merging/2.11.1 release.

File: pom.xml [hunk 3]
```diff
@@ -77,6 +77,11 @@
         </dependency>
         <dependency>
             <groupId>org.jenkins-ci.plugins.workflow</groupId>
+            <artifactId>workflow-api</artifactId>
+            <version>2.15</version>
```
Comment (jglick) [inline pom.xml]:
```
java.io.IOException: Pipeline: API v2.15 failed to load.
 - Pipeline: Step API v2.9 is older than required. To fix, install v2.10 or later.
```

Comment (abayer) [inline pom.xml]: Fixing.

**Pull request merged**